(WIP) Gtfs migration

TonyIlersich opened the pull request:
TTC does not fill the location_type field on stops, so they are interpreted as platforms, and we have no concept of stations. We may be able to infer stations from platform names or shared route or something like that.

Will need to install PostgreSQL on the server machine for this to work.

TonyIlersich committed:
ignore node_modules

TonyIlersich committed:
ignore env and build

TonyIlersich committed:
remove cached node_modules

TonyIlersich committed:
multiple changes
* add prisma, typescript, unzipper
* add GTFS typing
* create database schema
* create TtcApi wrapper class to load GTFS data and interface with db

TonyIlersich committed:
multiple changes
* implement api endpoints for subway routes and platforms
* add default colours for routes
* load and render subway routes and platforms on client side

TonyIlersich commented:
Rendering server side subway routes and platforms. We definitely want to consolidate into stations because all the stops appear duplicated now. You can see some of the visual weirdness here:
<img width="1052" height="705" alt="image" src="https://github.com/user-attachments/assets/a36cfce9-3995-499f-b250-e859ae1557ce" />
<img width="979" height="694" alt="image" src="https://github.com/user-attachments/assets/4b43f41a-7bca-4c0a-9ed3-4c42be349747" />
<img width="931" height="670" alt="image" src="https://github.com/user-attachments/assets/d453bced-0efe-4736-b841-30b00b445782" />

TonyIlersich commented:
We could smooth out the lines by using bezier curves instead of polylines. Leaflet has plugins for this:
- https://github.com/elfalem/Leaflet.curve
- https://github.com/slutske22/leaflet-spline

ailersic commented:
> We could smooth out the lines by using bezier curves instead of polylines. Leaflet has plugins for this:
> 
>     * https://github.com/elfalem/Leaflet.curve
> 
>     * https://github.com/slutske22/leaflet-spline

That's one option. We could also make the lines look smoother and more true to life using the coordinates in shapes.txt. Do you think that would substantially slow the site down?

ailersic commented:
> Rendering server side subway routes and platforms. We definitely want to consolidate into stations because all the stops appear duplicated now. You can see some of the visual weirdness here: [...]

Yeah that looks gross. When I hard-coded the stations initially, I took the midpoint of the platforms as the lat/long coordinates. The best bet is probably to automate that process with some special handling for transfer stations and Spadina.

TonyIlersich committed:
multiple changes
* add typing for GTFS realtime
* add shapes to backend
* pull GTFS alerts
* add alert endpoint
* display SignificantDelay alert segments on client
* generate inferred stations from platform names
* add subway station endpoint
* display stations instead of platforms on client
* switch GTFS schedule endpoint to "merged" dataset (this one has route colors! still no parent_station_id though)
* add case correction to platform names
* temporarily disabled other alert display logic

TonyIlersich committed:
turn off forceReload

TonyIlersich commented:
<img width="846" height="636" alt="image" src="https://github.com/user-attachments/assets/0aef4fb9-0304-486d-8def-c6de20d1627e" />

TonyIlersich commented:
> > Rendering server side subway routes and platforms. We definitely want to consolidate into stations because all the stops appear duplicated now. You can see some of the visual weirdness here: [...]
> 
> Yeah that looks gross. When I hard-coded the stations initially, I took the midpoint of the platforms as the lat/long coordinates. The best bet is probably to automate that process with some special handling for transfer stations and Spadina.

Added this now. Wasn't sure how to name the two Spadinas, so I went with "... - Line 1" and "... - Line 2". Also our system would naturally separate "Bloor Station" and "Yonge Station", so I added a special case to put them together, but we could split them up again if we want.

TonyIlersich commented:
> > We could smooth out the lines by using bezier curves instead of polylines. Leaflet has plugins for this:
> > ```
> > * https://github.com/elfalem/Leaflet.curve
> > 
> > * https://github.com/slutske22/leaflet-spline
> > ```
> 
> That's one option. We could also make the lines look smoother and more true to life using the coordinates in shapes.txt. Do you think that would substantially slow the site down?

I added the shapes in the backend now, about a million rows, we can probably condense it to a smaller format for the client, and of course most of that is busses. Will have to try and see how slow it is.

TonyIlersich committed:
multiple changes
* optimized subway endpoints
* render shapes on client
* link shapes and stops directly to routes
* improvements to station generation
* fix spadina tunnel

TonyIlersich commented:
Looking pretty good and no speed concerns about the shapes. The map looks like this right now:
<img width="1002" height="651" alt="image" src="https://github.com/user-attachments/assets/ea2014ed-28a3-4b17-85ef-04923fd19ba0" />

Spadina, Bloor-Yonge, and Union have position issues that I haven't fully diagnosed:
<img width="654" height="369" alt="image" src="https://github.com/user-attachments/assets/dff2097a-0ed0-41aa-acd4-148864c2d233" />
<img width="464" height="249" alt="image" src="https://github.com/user-attachments/assets/bc993a5e-fdfa-4540-ae47-da630c7fb62d" />
<img width="492" height="394" alt="image" src="https://github.com/user-attachments/assets/94b53f08-6156-4097-a5f8-d94fd1ae2ba8" />

We will also have to figure out how to line up the alerts with the shapes:
<img width="416" height="485" alt="image" src="https://github.com/user-attachments/assets/6e05085a-ec48-4818-8d47-8b67569ee385" />

ailersic commented:
Those shapes are looking great. Maybe for the station position issues, we can "snap" the station to the nearest point on the shape for its line, or the nearest intersection point of both lines for transfer stations? With each station assigned to a point along the shape, the alert lines can follow the shape coordinates between station i and station j.

TonyIlersich commented:
> Those shapes are looking great. Maybe for the station position issues, we can "snap" the station to the nearest point on the shape for its line, or the nearest intersection point of both lines for transfer stations? With each station assigned to a point along the shape, the alert lines can follow the shape coordinates between station i and station j.

The stations aren't linked to the shape right now, so it will take some work to massage the data. Stations are snapping to the nearest point defined on the shape, but for some reason those three aren't quite making it. It might be because they have multiple shapes that dont quite intersect, but at least for bloor-yonge they clearly do, so I suspect we have to lerp between shape points to find the true snap position. I can try these next time.

TonyIlersich committed:
multiple changes
* link stations to shapes
* produce route shape segments between stations for client
* render alerts on route shape segments instead of straight-line
* switch subway route semantics from type=SubwayMetro to id list (to includ LRT)
* improvements to station-shape snapping
* add geometry utils

TonyIlersich commented:
<img width="817" height="662" alt="image" src="https://github.com/user-attachments/assets/0d83eee0-64f3-4593-b837-a5d5d348ec8d" />

TonyIlersich commented:
I'm going to pause on this for now so you have time to review the changes.

TonyIlersich committed:
multiple changes:
* remove Trip and TripStop
* optimize shapes
* remove alert cause

TonyIlersich committed:
smooth shapes

TonyIlersich committed:
merge main
* adapt endpoints for serverless format
* pregenerate protobuf types to avoid runtime code generation
* reconfigure prisma for d1
* move static assets into public folder
* made TtcApi wrapper class (mostly) stateless

TonyIlersich committed:
rewrite build commands for cloudflare deployment

TonyIlersich committed:
commit env.local

TonyIlersich committed:
remove local env

TonyIlersich committed:
fix pages configuration

TonyIlersich committed:
fix prisma datasource and migrations

TonyIlersich committed:
wip migration

TonyIlersich committed:
wip migration

TonyIlersich committed:
Merge branch 'gtfs-migration' of https://github.com/ailersic/ttc-service-map into gtfs-migration

TonyIlersich committed:
fix typo

TonyIlersich committed:
wip: bind api token?

TonyIlersich committed:
wip: fix secret command

TonyIlersich committed:
wip: don't need secret in pipeline?

TonyIlersich committed:
remove build:remote

TonyIlersich committed:
wip: remove datasource

TonyIlersich committed:
up cpu time

TonyIlersich committed:
add script to seed database

TonyIlersich committed:
add db seed script

TonyIlersich committed:
add benchmarking to TtcApi endpoint functions

TonyIlersich committed:
benchmark client side loading and decouple map from alerts

TonyIlersich committed:
WIP troubleshooting cloudflare logging

TonyIlersich committed:
WIP

TonyIlersich committed:
switch to subway-specific alert endpoint to improve initial load

TonyIlersich committed:
fix date

ailersic committed:
added icons, highlights, tooltips

ailersic committed:
started routes legend

ailersic committed:
made invisible routes slightly less transparent

ailersic committed:
added route categories

ailersic committed:
updated station name

ailersic committed:
updated station name

TonyIlersich committed:
shape inferral for shapeless routes, auto-upload to d1, name/casing fixes, formerly field on stations

TonyIlersich committed:
Merge branch 'gtfs-migration' of https://github.com/ailersic/ttc-service-map into gtfs-migration

TonyIlersich committed:
ideas for spadina

TonyIlersich committed:
null check

TonyIlersich committed:
troubleshooting

TonyIlersich committed:
Merge branch 'main' of https://github.com/ailersic/ttc-service-map into gtfs-migration

TonyIlersich committed:
add formerly to station tooltip

ailersic committed:
fixed start/end station names in alerts
